default/AI/ResearchAI.py

```diff
@@ -20,21 +32,402 @@ def get_research_index():
+    empire = fo.getEmpire()
+    research_reqs.clear()
+
+    completed_techs = get_completed_techs()
+    for tech_name in fo.techs():
+        if tech_is_complete(tech_name):
+            research_reqs[tech_name] = ([], 0, 0, 0)
+            continue
+        this_tech = fo.getTech(tech_name)
+        prereqs = [preReq for preReq in this_tech.recursivePrerequisites(empire.empireID) if preReq not in completed_techs]
+        base_cost = this_tech.researchCost(empire.empireID)
+        progress = empire.researchProgress(tech_name)
+        cost = max(0.0, base_cost - progress)
+        proportion_remaining = cost / max(base_cost, 1.0)
+        this_time = this_tech.researchTime(fo.empireID())
+        turns_needed = max(1, math.ceil(proportion_remaining * this_time))  # even if fully paid needs one turn
+        per_turn_cost = float(base_cost) / max(1.0, this_time)
+
+        # TODO: the following timing calc treats prereqs as inherently sequential; consider in parallel when able
+        for prereq in prereqs:
+            prereq_tech = fo.getTech(prereq)
+            if not prereq_tech:
+                continue
+            base_cost = prereq_tech.researchCost(empire.empireID)
+            progress = empire.researchProgress(prereq)
+            prereq_cost = max(0.0, base_cost - progress)
+            proportion_remaining = prereq_cost / max(base_cost, 1.0)
+            this_time = prereq_tech.researchTime(fo.empireID())
+            turns_needed += max(1, math.ceil(proportion_remaining * this_time))
+            cost += prereq_cost
+        research_reqs[tech_name] = (prereqs, cost, turns_needed, per_turn_cost)
+
+def tech_cost_sort_key(tech_name):
+    return research_reqs.get(tech_name, ([],0,0,0))[REQS_COST_IDX]
 
-def exclude_tech(tech_name):
-    return ((foAI.foAIstate.aggression < AIDependencies.TECH_EXCLUSION_MAP_1.get(tech_name, fo.aggression.invalid)) or
-            (foAI.foAIstate.aggression > AIDependencies.TECH_EXCLUSION_MAP_2.get(tech_name, fo.aggression.maniacal)) or
-            tech_name in TechsListsAI.unusable_techs())
-            
+def tech_time_sort_key(tech_name):
+    return research_reqs.get(tech_name, ([],0,0,0))[REQS_TIME_IDX]
 
 def generate_research_orders():
     """generate research orders"""
+
+    # initializing priority functions here within generate_research_orders() to avoid import race
+
+    DEFENSIVE = foAI.foAIstate.aggression <= fo.aggression.cautious
+
+    # keys are "PREFIX1", as in "DEF" or "SPY"
+    if not primary_prefix_priority_funcs:
+        primary_prefix_priority_funcs.update({
+            AIDependencies.DEFENSE_TECHS_PREFIX: partial(const_priority, 2.0) if DEFENSIVE else partial(if_enemies, 0.2)
+            })
+
+    # keys are "PREFIX1_PREFIX2", as in "SHP_WEAPON"
+    if not secondary_prefix_priority_funcs:
+        secondary_prefix_priority_funcs.update({
+            AIDependencies.WEAPON_PREFIX: partial(ship_usefulness, partial(if_enemies, 0.2), MIL_IDX)
+            })
```

Comment:
Instead of having multiple dicts like this, we could also once loop over all techs, build a list of techs with the corresponding prefix(es) and use that list (converted to a tuple) as key in the ```misc_group_priority_funcs``` dict.
That way only one dict handles all the function calls. Seems somewhat cleaner to me.